tools/IoTEdgeDevOps/DevOpsLib/AzureActiveDirectory.cs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace DevOpsLib
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Net.Http;
+    using System.Net.Http.Headers;
+    using System.Reflection.Metadata.Ecma335;
+    using System.Threading.Tasks;
+    using Newtonsoft.Json.Linq;
+
+    public sealed class AzureActiveDirectory
+    {
+        string azureActiveDirTenant = null;
+        string azureActiveDirClientId = null;
+        string azureActiveDirClientSecret = null;
+
+        string azureResource = null;
+        string accessToken = null;
+        DateTime accessTokenExpiration = new DateTime(DateTime.MinValue.Ticks);
+        readonly object locker = new object();
+
+        public AzureActiveDirectory(
+            string azureActiveDirTenant,
+            string azureActiveDirClientId,
+            string azureActiveDirClientSecret)
+        {
+            this.azureActiveDirTenant = azureActiveDirTenant;
+            this.azureActiveDirClientId = azureActiveDirClientId;
+            this.azureActiveDirClientSecret = azureActiveDirClientSecret;
+        }
+
+        // GetAccessToken(4) is required to be called before invoking GetAccessToken(0), or GetAccessToken(1)
+        // Trigger Azure Active Directory (AAD) for an OAuth2 client credential for an azure resource access.
+        // API reference: https://dev.loganalytics.io/documentation/Authorization/OAuth2
+        public async Task<string> GetAccessToken(string azureResource)
+        {
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirTenant, nameof(azureActiveDirTenant));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientId, nameof(azureActiveDirClientId));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientSecret, nameof(azureActiveDirClientSecret));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureResource, nameof(azureResource));
+
+            try
+            {
+                
+                if (this.IsAccessTokenExpired() ||
```

Comment:
nit: wrap all these condition in 1 method named "RequireAccessToken" or "RefreshAccessToken".